pkg/schemadsl/lexer/lex.go

```diff
@@ -27,16 +28,46 @@ func createLexer(source input.InputSource, input string) *Lexer {
 
 // run runs the state machine for the lexer.
 func (l *Lexer) run() {
-	for l.state = lexSource; l.state != nil; {
-		l.state = l.state(l)
+	defer func() {
+		close(l.tokens)
+	}()
+	l.withLock(func() {
+		l.state = lexSource
+	})
+	var state stateFn
+	for {
+		l.withRLock(func() {
+			state = l.state
+		})
+		if state == nil {
+			break
+		}
+		l.withLock(func() {
+			l.state = state(l)
```

Comment:
This will lock during the entire lex; you probably want to call the `state(l)` outside of the lock

Reply:
Good call - updated